### PR TITLE
feat: TrustlessGatewayBlockBroker prioritizes & tries all gateways

### PR DESCRIPTION
**Pull request by SgtPooki**
Slight changes on top of #280

- keep track of individual gateway status and expose reliability rating we can use to sort/prioritize
- Try all gateways in TrustlessGatewayBlockBroker, sequentially, sorted by reliability

### todo

- [x] fix tests (hasher)
- [x] Add tests for racing block providers
- [x] Add tests for prioritizion of trustlessGateways that error/succeed or are untried

### CODE REVIEW

File: packages/helia/src/block-providers/trustless-gateway-block-provider.ts [hunk 1]
```diff
@@ -6,78 +6,143 @@ import type { ProgressEvent, ProgressOptions } from 'progress-events'
...
+      return 1
+    }
 
-      throw err
+    // The gateway has > 0 attempts; If we have never encountered an error, consider it 100% reliable.
+    // Even if a gateway has no successes, it is still considered more reliable than a gateway with errors,
+    // because it may have been used and aborted, or beaten by another BlockProvider.
+    if (this.#errors === 0) {
+      return 1
     }
```
Comment (SgtPooki) [inline packages/helia/src/block-providers/trustless-gateway-block-provider.ts:104]: We need to remove this in order to ensure slow gateways that never error are deprioritized properly.

**Comment by achingbrain** [comment]
Can we scope this PR to improving the `TrustlessGatewayProvider`?  I think all the other points can be taken care of in #280

File: packages/helia/src/block-providers/trustless-gateway-block-provider.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+> {
+  private readonly gateways: TrustlessGateway[]
+
+  constructor (urls: Array<string | URL>) {
+    this.gateways = urls.map((url) => new TrustlessGateway(url))
+  }
+
+  async get (cid: CID, options: AbortOptions & ProgressOptions<TrustlessGatewayGetBlockProgressEvents> = {}): Promise<Uint8Array> {
+    // Loop through the gateways until we get a block or run out of gateways
+    for (const gateway of this.gateways.sort((a, b) => b.reliability - a.reliability)) {
```
Comment (SgtPooki) [inline packages/helia/src/block-providers/trustless-gateway-block-provider.ts:135]:
```suggestion
	const sortedGateways = this.gateways.sort((a, b) => b.reliability - a.reliability)
    for (const gateway of sortedGateways) {
```

File: packages/helia/src/block-providers/trustless-gateway-block-provider.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        const block = await gateway.getRawBlock(cid, options.signal)
+        log('got block for %c from %s', cid, gateway.url)
+
+        return block
+      } catch (err) {
+        log.error('failed to get block for %c from %s', cid, gateway.url, err)
+      }
+    }
+
+    throw new Error(`unable to fetch raw block for CID ${cid} from any gateway`)
```
Comment (SgtPooki) [inline packages/helia/src/block-providers/trustless-gateway-block-provider.ts:147]:
```suggestion
    throw new CodeError(`unable to fetch raw block for CID ${cid} from any gateway`, ADD_NEW_GATEWAY_ERROR)
```

Comment (SgtPooki) [inline packages/helia/src/block-providers/trustless-gateway-block-provider.ts:147]:
```suggestion
      } catch (err) {
      	aggErrors.push(err)
        log.error('failed to get block for %c from %s', cid, gateway.url, err)
      }
    }
    

    throw new AggregateError(aggErrors, `unable to fetch raw block for CID ${cid} from any gateway`)
```

**Pull request merged**